packages/grid/_modules_/grid/components/toolbar/DensityPicker.tsx

```diff
@@ -0,0 +1,104 @@
+import * as React from 'react';
+import Button from '@material-ui/core/Button';
+import Menu from '@material-ui/core/Menu';
+import MenuItem from '@material-ui/core/MenuItem';
+import ListItemIcon from '@material-ui/core/ListItemIcon';
+import ListItemText from '@material-ui/core/ListItemText';
+import { useIcons } from '../../hooks/utils/useIcons';
+import { ApiContext } from '../api-context';
+import { DensityTypes, Density } from '../../models/gridOptions';
+import { useGridSelector } from '../../hooks/features/core/useGridSelector';
+import { optionsSelector } from '../../hooks/utils/useOptionsProp';
+import { DensityOption } from '../../models/api/densityApi';
+import { densitySizeSelector } from '../../hooks/features/density';
+
+export const DensityPicker = React.memo(function DensityPicker() {
+  const apiRef = React.useContext(ApiContext);
+  const { density, rowHeight, headerHeight } = useGridSelector(apiRef, optionsSelector);
+  const densitySize = useGridSelector(apiRef, densitySizeSelector);
+  const [anchorEl, setAnchorEl] = React.useState(null);
+  const icons = useIcons();
+
+  const DensitySmallIcon = icons!.DensitySmall!;
+  const DensityMediumIcon = icons!.DensityMedium!;
+  const DensityLargeIcon = icons!.DensityLarge!;
+
+  const DensityOptions: Array<DensityOption> = [
+    {
+      icon: <DensitySmallIcon />,
+      label: DensityTypes.Small,
+    },
+    {
+      icon: <DensityMediumIcon />,
+      label: DensityTypes.Medium,
+    },
+    {
+      icon: <DensityLargeIcon />,
+      label: DensityTypes.Large,
+    },
+  ];
+
+  React.useEffect(() => {
+    apiRef!.current.setDensity(density, headerHeight, rowHeight);
+  }, [apiRef, density, rowHeight, headerHeight]);
+
+  const getSelectedDensityIcon = React.useCallback((): React.ReactElement => {
```

Comment:
you could use a key value object and avoid the switch and the densityOptions block

Reply:
I could but I also need to loop through the object to create the menu options and currently there is a problem with Typescript that if you have a variable that is of a custom type (from an Interface) then you need to add an iterator function to that interface handle the `for of` loop. I'm open to suggestions.